File: ports/nim/src/watch.nim
## watch.nim - sprite-tool watch subcommand
## Poll a sprite's beads tracker task for progress.

import os, osproc, strutils, times
import sprite

const watchUsage* = """
Usage: sprite-tool watch <sprite-name> [task-id] [poll-interval]

Arguments:
  sprite-name     Name of the sprite to watch
  task-id         Beads task ID to track (default: auto-detect first open critical task)
  poll-interval   Seconds between polls (default: 30)

Examples:
  sprite-tool watch ember-red-hawk
  sprite-tool watch ember-red-hawk CRM-1
  sprite-tool watch ember-red-hawk CRM-1 60
"""

proc runWatch*(args: seq[string]) =
  if args.len < 1:
    echo watchUsage
    quit(1)

  let spriteName = args[0]
  var taskId = if args.len > 1: args[1] else: ""
  var pollInterval = 30
  if args.len > 2:
    try:
      pollInterval = parseInt(args[2])
    except ValueError:
      echo "ERROR: Invalid poll interval: " & args[2]
      quit(1)

  # Auto-detect tracker task if not specified
  if taskId == "":
    echo "Detecting tracker task..."
    let (critOut, critRc) = sxCapture(spriteName,
      "cd /home/sprite && bd list --priority critical 2>/dev/null | head -1 | awk '{print $1}'")
    if critRc == 0 and critOut.strip() != "":
      taskId = critOut.strip()
    else:
      echo "No critical task found. Falling back to first open task..."
      let (openOut, openRc) = sxCapture(spriteName,
        "cd /home/sprite && bd list 2>/dev/null | head -1 | awk '{print $1}'")
      if openRc == 0 and openOut.strip() != "":
        taskId = openOut.strip()

    if taskId == "":
      echo "ERROR: No beads tasks found on sprite '" & spriteName & "'"
      echo "Specify a task ID manually: sprite-tool watch " & spriteName & " <task-id>"
      quit(1)

    echo "Tracking task: " & taskId

  echo "Watching sprite '" & spriteName & "' task '" & taskId & "' (every " & $pollInterval & "s)"
  echo "Press Ctrl+C to stop"
  echo ""

  while true:
    # Clear screen
    discard execCmd("clear")

    let timeStr = now().format("HH:mm:ss")
    echo "=== sprite-watch: " & spriteName & " / " & taskId & " === " & timeStr & " ==="
    echo ""

    # Show task status
    let showRc = sx(spriteName, "cd /home/sprite && bd show " & taskId & " 2>/dev/null")
    if showRc != 0:
      echo "(could not read task)"
    echo ""

    # Show recent comments
    echo "--- Recent updates ---"
    let commentsRc = sx(spriteName, "cd /home/sprite && bd comments " & taskId & " 2>/dev/null | tail -8")
    if commentsRc != 0:
      echo "(no comments)"
    echo ""

    # Check if done
    let (statusOut, _) = sxCapture(spriteName,
      "cd /home/sprite && bd show " & taskId & " 2>/dev/null | grep -i status")
    let statusLower = statusOut.toLowerAscii()
    if statusLower.contains("closed") or statusLower.contains("done") or statusLower.contains("completed"):
      echo "=========================================="
      echo "PROJECT COMPLETE"
      echo "=========================================="
      break

    sleep(pollInterval * 1000)

File: ports/fsharp/Watch.fs
module SpriteTool.Watch

open System
open System.Threading

let private usage () =
    printfn
        "Usage: sprite-tool watch <sprite-name> [task-id] [poll-interval]\n\
         \n\
         Arguments:\n\
         \  sprite-name     Name of the sprite to watch\n\
         \  task-id         Beads task ID to track (default: auto-detect first open critical task)\n\
         \  poll-interval   Seconds between polls (default: 30)\n\
         \n\
         Examples:\n\
         \  sprite-tool watch ember-red-hawk\n\
         \  sprite-tool watch ember-red-hawk CRM-1\n\
         \  sprite-tool watch ember-red-hawk CRM-1 60"

    exit 1

/// Run command inside sprite, capture stdout (suppresses stderr).
let private sxCapture (spriteName: string) (cmd: string) =
    Sprite.sx spriteName cmd false
    |> Option.defaultValue ""

/// Run command inside sprite, print output, return it.
let private sxPrint (spriteName: string) (cmd: string) =
    let output = sxCapture spriteName cmd

    if not (String.IsNullOrEmpty output) then
        printfn "%s" output

    output

let run (args: string list) =
    if args.Length < 1 then
        usage ()

    let spriteName = args.[0]

    let mutable taskId =
        if args.Length > 1 then args.[1] else ""

    let pollInterval =
        if args.Length > 2 then
            match Int32.TryParse args.[2] with
            | true, v -> v
            | false, _ ->
                eprintfn "Error: invalid poll-interval '%s' (must be integer)" args.[2]
                exit 1
        else
            30

    // Auto-detect tracker task if not specified
    if String.IsNullOrEmpty taskId then
        printfn "Detecting tracker task..."

        taskId <-
            sxCapture
                spriteName
                "cd /home/sprite && bd list --priority critical 2>/dev/null | head -1 | awk '{print $1}'"

        if String.IsNullOrEmpty taskId then
            printfn "No critical task found. Falling back to first open task..."

            taskId <-
                sxCapture spriteName "cd /home/sprite && bd list 2>/dev/null | head -1 | awk '{print $1}'"

        if String.IsNullOrEmpty taskId then
            eprintfn "ERROR: No beads tasks found on sprite '%s'" spriteName
            eprintfn "Specify a task ID manually: sprite-tool watch %s <task-id>" spriteName
            exit 1

        printfn "Tracking task: %s" taskId

    printfn "Watching sprite '%s' task '%s' (every %ds)" spriteName taskId pollInterval
    printfn "Press Ctrl+C to stop"
    printfn ""

    let cts = new CancellationTokenSource()

    Console.CancelKeyPress.Add(fun evtArgs ->
        evtArgs.Cancel <- true
        cts.Cancel())

    try
        while not cts.Token.IsCancellationRequested do
            // Clear screen using ANSI escape
            printf "\x1B[2J\x1B[H"
            Console.Out.Flush()

            let now = DateTime.Now.ToString("HH:mm:ss")
            printfn "=== sprite-watch: %s / %s === %s ===" spriteName taskId now
            printfn ""

            // Show task status
            let taskOutput =
                sxPrint spriteName (sprintf "cd /home/sprite && bd show %s 2>/dev/null" taskId)

            if String.IsNullOrEmpty taskOutput then
                printfn "(could not read task)"

            printfn ""

            // Show recent comments
            printfn "--- Recent updates ---"

            let commentsOutput =
                sxPrint spriteName (sprintf "cd /home/sprite && bd comments %s 2>/dev/null | tail -8" taskId)

            if String.IsNullOrEmpty commentsOutput then
                printfn "(no comments)"

            printfn ""

            // Check if done
            let status =
                sxCapture spriteName (sprintf "cd /home/sprite && bd show %s 2>/dev/null | grep -i status" taskId)

            if not (String.IsNullOrEmpty status) then
                let statusLower = status.ToLowerInvariant()

                if
                    statusLower.Contains "closed"
                    || statusLower.Contains "done"
                    || statusLower.Contains "completed"
                then
                    printfn "=========================================="
                    printfn "PROJECT COMPLETE"
                    printfn "=========================================="
                    cts.Cancel()
                else
                    Thread.Sleep(pollInterval * 1000)
            else
                Thread.Sleep(pollInterval * 1000)
    with :? OperationCanceledException ->
        ()

    printfn ""
    printfn "Stopped."
